repo.go

```diff
@@ -477,13 +524,150 @@ func (r *Repo) RevokeKeyWithExpires(keyRole, id string, expires time.Time) error
 		root.Version++
 	}
 
-	err = r.setTopLevelMeta("root.json", root)
+	err = r.setMeta("root.json", root)
 	if err == nil {
 		fmt.Println("Revoked", keyRole, "key with ID", id, "in root metadata")
 	}
 	return err
 }
 
+// AddDelegatedRole is equivalent to AddDelegatedRoleWithExpires, but
+// with a default expiration time.
+func (r *Repo) AddDelegatedRole(delegator string, delegatedRole data.DelegatedRole, keys []*data.PublicKey) error {
+	return r.AddDelegatedRoleWithExpires(delegator, delegatedRole, keys, data.DefaultExpires("targets"))
+}
+
+// AddDelegatedRoleWithExpires adds a delegation from the delegator to the
+// role specified in the role argument. Key IDs referenced in role.KeyIDs
+// should have corresponding Key entries in the keys argument. New metadata is
+// written with the given expiration time.
+func (r *Repo) AddDelegatedRoleWithExpires(delegator string, delegatedRole data.DelegatedRole, keys []*data.PublicKey, expires time.Time) error {
+	expires = expires.Round(time.Second)
+
+	t, err := r.targets(delegator)
+	if err != nil {
+		return fmt.Errorf("error getting delegator (%q) metadata: %w", delegator, err)
+	}
+
+	if t.Delegations == nil {
+		t.Delegations = &data.Delegations{}
+		t.Delegations.Keys = make(map[string]*data.PublicKey)
+	}
+
+	for _, keyID := range delegatedRole.KeyIDs {
+		for _, key := range keys {
+			if key.ContainsID(keyID) {
+				t.Delegations.Keys[keyID] = key
+				break
+			}
+		}
+	}
+
+	for _, r := range t.Delegations.Roles {
+		if r.Name == delegatedRole.Name {
+			return fmt.Errorf("role: %s is already delegated to by %s", delegatedRole.Name, r.Name)
+		}
+	}
+	t.Delegations.Roles = append(t.Delegations.Roles, delegatedRole)
+	t.Expires = expires
+
+	delegatorFile := delegator + ".json"
+	if !r.local.FileIsStaged(delegatorFile) {
+		t.Version++
+	}
+
+	err = r.setMeta(delegatorFile, t)
+	if err != nil {
+		return fmt.Errorf("error setting metadata for %q: %w", delegatorFile, err)
+	}
+
+	delegatee := delegatedRole.Name
+	dt, err := r.targets(delegatee)
+	if err != nil {
+		return fmt.Errorf("error getting delegatee (%q) metadata: %w", delegatee, err)
+	}
+	dt.Expires = expires
+
+	delegateeFile := delegatee + ".json"
+	if !r.local.FileIsStaged(delegateeFile) {
```

Comment:
Do we expect this file to possibly already be created? On one hand I suppose this could be used to "update" a delegation, but on the other hand, this function will also clear any targets that were in the delegated file.

Side question: Does the version of the delegateeFile bump if the signers change, but the content (targets and further delegations) do not?

Reply:
> Do we expect this file to possibly already be created?

`delegateeFile` may already be created. For example, if you have a diamond delegation graph with edges (A->B), (A->C), (B->D), (C->D), `D.json` will already exist when you create the delegation from C to D.

> this function will also clear any targets that were in the delegated file

This function should not clear any targets in `delegateeFile`, since the call `r.targets(delegatee)` several lines up returns existing metadata if it exists, or creates empty targets metadata if necessary. I added a test in https://github.com/theupdateframework/go-tuf/pull/175/commits/be25f9324374ffb1c599d3f801150cfe6a6aa344 to verify this.

> Does the version of the delegateeFile bump if the signers change, but the content (targets and further delegations) do not?

The version number should not change unless the content within the `signed` object changes. Adding or removing from the `signatures` section of a `data.Signed` meta file should not change the version number. However, `AddDelegatedRoleWithExpires` does currently refresh the expiration of the delegatee file (defined inside `signed`), which will trigger a version bump (also verified in https://github.com/theupdateframework/go-tuf/pull/175/commits/be25f9324374ffb1c599d3f801150cfe6a6aa344).